cores/esp32/IPAddress.cpp

```diff
@@ -94,151 +99,20 @@ IPAddress::IPAddress(const char *address)
-    if (_type == IPv4) {
-        return _address.bytes[IPADDRESS_V4_BYTES_INDEX + index];
-    }
-    return _address.bytes[index];
-}
-
-size_t IPAddress::printTo(Print& p) const
-{
-    size_t n = 0;
-
-    if (_type == IPv6) {
-        // IPv6 IETF canonical format: compress left-most longest run of two or more zero fields, lower case
-        int8_t longest_start = -1;
-        int8_t longest_length = 1;
-        int8_t current_start = -1;
-        int8_t current_length = 0;
-        for (int8_t f = 0; f < 8; f++) {
-            if (_address.bytes[f * 2] == 0 && _address.bytes[f * 2 + 1] == 0) {
-                if (current_start == -1) {
-                    current_start = f;
-                    current_length = 1;
-                } else {
-                    current_length++;
-                }
-                if (current_length > longest_length) {
-                    longest_start = current_start;
-                    longest_length = current_length;
-                }
-            } else {
-                current_start = -1;
-            }
-        }
-        for (int f = 0; f < 8; f++) {
-            if (f < longest_start || f >= longest_start + longest_length) {
-                uint8_t c1 = _address.bytes[f * 2] >> 4;
-                uint8_t c2 = _address.bytes[f * 2] & 0xf;
-                uint8_t c3 = _address.bytes[f * 2 + 1] >> 4;
-                uint8_t c4 = _address.bytes[f * 2 + 1] & 0xf;
-                if (c1 > 0) {
-                    n += p.print((char)(c1 < 10 ? '0' + c1 : 'a' + c1 - 10));
-                }
-                if (c1 > 0 || c2 > 0) {
-                    n += p.print((char)(c2 < 10 ? '0' + c2 : 'a' + c2 - 10));
-                }
-                if (c1 > 0 || c2 > 0 || c3 > 0) {
-                    n += p.print((char)(c3 < 10 ? '0' + c3 : 'a' + c3 - 10));
-                }
-                n += p.print((char)(c4 < 10 ? '0' + c4 : 'a' + c4 - 10));
-                if (f < 7) {
-                    n += p.print(':');
-                }
-            } else if (f == longest_start) {
-                if (longest_start == 0) {
-                    n += p.print(':');
-                }
-                n += p.print(':');
-            }
-        }
-        return n;
-    }
-
-    // IPv4
-    for (int i =0; i < 3; i++)
-    {
-        n += p.print(_address.bytes[IPADDRESS_V4_BYTES_INDEX + i], DEC);
-        n += p.print('.');
-    }
-    n += p.print(_address.bytes[IPADDRESS_V4_BYTES_INDEX + 3], DEC);
-    return n;
-}
-
-String IPAddress::toString4() const
-{
-    char szRet[16];
-    snprintf(szRet, sizeof(szRet), "%u.%u.%u.%u", _address.bytes[IPADDRESS_V4_BYTES_INDEX], _address.bytes[IPADDRESS_V4_BYTES_INDEX + 1], _address.bytes[IPADDRESS_V4_BYTES_INDEX + 2], _address.bytes[IPADDRESS_V4_BYTES_INDEX + 3]);
-    return String(szRet);
-}
-
-String IPAddress::toString6() const
+String IPAddress::toString() const
 {
     StreamString s;
-    s.reserve(40);
     printTo(s);
```

Comment:
Glad you kept the RFC canonical printTo() algorithm ... at one point it looked like you had a fixed length format.

Note: I originally had a single toString() function, but added toString4() and toString6() to be compatible with changes in ArduinoCore (although they were only protected, so not that important).

If you want to remove them, that is fine (cleaner, I think) ... but be aware it is drifting from the common (but the public API is the same, so I think that is okay).

Reply:
Originally Arduino.cc was printing fixed width. I changed the method to use `printTo` instead, in order to print correct format

Reply:
Ah yes, I saw that change and raised an issue for that https://github.com/arduino/ArduinoCore-API/issues/228

Somebody made that change since I wrote the original printTo() (and it was pretty bad, to put in a fixed width toString() when printTo() already had the correct algorithm).

Originally ArduinoCore only had printTo(), so I only implemented the algorithm there. Arduino-ESP32 had a toString() as well, so I just used printTo() to fill the string buffer. I kept the same order in the class though in my recent merged PR https://github.com/espressif/arduino-esp32/pull/7174

When I updated for recent changes in ArduinoCore, I added toString4() and toString6() to match, but did not copy the implementation of toString6() that someone else added, because I thought it was just bad code (because printTo() already existed).

But, it is fine to use the code straight from core... all except for that bad implementation of toString6(), so I'm glad you changed that!